### PR TITLE
Fee calculation using the cli

### PR DESCRIPTION
**Pull request by gege251**
Solves: #10

I tried to reuse functions from the cardano-node repository, but had a few problems getting the required values for the computation (namely EraHistory and SystemStart), so I decided to use the CLI for this.

I had to add the following steps to the tx submit flow:
1. calculating minimum utxo
2. build a tx with build-raw, this doesnt include any fees yet, its only purpose is to be used in the following step
3. calculate minimum fees, this uses a tx file as its input

For the second step, I had to use the raw transaction build mode, so the cli wouldn't worry about balancing, and would create a tx even if there aren't enough inputs. However, I had to introduce execution unit calculation, which is needed by the cli in raw mode.

### CODE REVIEW

File: src/MLabsPAB/PreBalance.hs [hunk 1]
```diff
@@ -160,9 +195,9 @@ balanceNonAdaOuts ownPkh utxos tx =
             txOuts
           (txOut@TxOut {txOutValue = v} : txOuts, txOuts') ->
             txOut {txOutValue = v <> nonAdaChange} : (txOuts <> txOuts')
-   in if Value.isZero nonAdaChange
-        then tx
-        else tx {txOutputs = outputs}
+   in if isValueNat nonAdaChange
+        then if Value.isZero nonAdaChange then Right tx else Right $ tx {txOutputs = outputs}
```
Comment (samuelWilliams99) [inline src/MLabsPAB/PreBalance.hs:199]:
```suggestion
        then Right $ if Value.isZero nonAdaChange then tx else tx {txOutputs = outputs}
```

**Review by samuelWilliams99 — COMMENTED** [review]
Impressive stuff!

File: src/MLabsPAB/CardanoCLI.hs [hunk 2]
```diff
@@ -97,34 +114,92 @@ utxosAt pabConf address = do
...
+  callCommand
+    ShellArgs
+      { cmdName = "cardano-cli"
+      , cmdArgs =
+          mconcat
+            [ ["transaction", "calculate-min-fee"]
+            , ["--tx-body-file", "tx.raw"]
+            , ["--tx-in-count", showText $ 1 + length (txInputs unBalancedTxTx)]
+            , ["--tx-out-count", showText $ length $ txOutputs unBalancedTxTx]
+            , ["--witness-count", showText $ 1 + length (Map.keys unBalancedTxRequiredSignatories)]
```
[the remark this answers is not in the document]
Comment (gege251) [reply src/MLabsPAB/CardanoCLI.hs:153]: At first, I didn't intend to use a prebalanced tx, so I added one txinput and one witness, and unfortunately forgot about them later.. I'll remove it now.

File: src/MLabsPAB/PreBalance.hs (hunk 1 above)
Comment (gege251) [reply src/MLabsPAB/PreBalance.hs]: Addressed

**Pull request merged**